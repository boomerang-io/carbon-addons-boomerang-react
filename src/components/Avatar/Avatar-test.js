import React from 'react';
import { shallow } from 'enzyme';

import Avatar from './Avatar';

describe('Default Avatar', () => {
  describe('Renders as expected', () => {
    const wrapper = shallow(<Avatar src="ibm.com/fake/path/to/img.png" />);
    it('size should be medium by default', () => {
      expect(wrapper.hasClass('bx--bmrg-avatar --medium')).toEqual(true);
    });
    it('should include child content', () => {
      expect(wrapper.find('img').length).toBe(1);
    });
  });
});
