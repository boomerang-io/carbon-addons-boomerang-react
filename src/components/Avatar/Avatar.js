import React, { useState } from 'react';
import PropTypes from 'prop-types';
import { settings } from 'carbon-components';

const { prefix } = settings;

const UserIcon = (props) => {
  return (
    <svg
      version="1.1"
      id="Layer_1"
      xmlns="http://www.w3.org/2000/svg"
      x="0px"
      y="0px"
      width="16px"
      height="16px"
      viewBox="0 0 16 16"
      {...props}
    >
      <title>user</title>
      <desc>Created with Sketch.</desc>
      <g id="user">
        <path
          id="Combined-Shape"
          fill="#f8fafa"
          d="M4,13.7C5.1,14.5,6.5,15,8,15s2.9-0.5,4-1.3v-1.2c0-0.8-0.7-1.5-1.5-1.5h-5C4.7,11,4,11.7,4,12.5V13.7
		z M3,12.9v-0.4C3,11.1,4.1,10,5.5,10h5c1.4,0,2.5,1.1,2.5,2.5v0.4c1.2-1.3,2-3,2-4.9c0-3.9-3.1-7-7-7S1,4.1,1,8
		C1,9.9,1.8,11.6,3,12.9z M8,16c-4.4,0-8-3.6-8-8s3.6-8,8-8s8,3.6,8,8S12.4,16,8,16z"
        />
        <path
          id="Oval-17"
          fill="#f8fafa"
          d="M8,8c1.1,0,2-0.9,2-2S9.1,4,8,4S6,4.9,6,6S6.9,8,8,8z M8,9C6.3,9,5,7.7,5,6s1.3-3,3-3s3,1.3,3,3S9.7,9,8,9z"
        />
      </g>
    </svg>
  );
};

const Avatar = ({ className, size, src, style, userName, ...rest }) => {
  const [error, setError] = useState(false);

  const altText = `Avatar for ${userName}`;

  return error || !src ? (
    <UserIcon alt={altText} className={`${className} --${size || ''}`} style={style} />
  ) : (
    <img
      alt={altText}
      className={`${className} --${size || ''}`}
      src={src}
      style={style}
      onError={() => setError(true)}
      {...rest}
    />
  );
};

Avatar.propTypes = {
  className: PropTypes.string,
  size: PropTypes.oneOf(['small', 'medium', 'large']),
  src: PropTypes.string.isRequired,
  style: PropTypes.object,
  userName: PropTypes.string,
};

Avatar.defaultProps = {
  className: `${prefix}--bmrg-avatar`,
  size: 'medium',
};

export default Avatar;
