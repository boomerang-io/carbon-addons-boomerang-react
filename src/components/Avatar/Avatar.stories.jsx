
import Component from "./Avatar";

export default {
  title: "Components/Avatar",
  components: Component,
  argTypes: {
    src: {
      description: "URL to avatar image"
    },
    size: {
      description: "Control the size",
      control: "select",
      options: ["small", "medium", "large"]
    },
    userName: {
      description: "Used for alt text"
    },
    style: {
      description: "Pass styles"
    },
    className: {
      description: "Pass custom class"
    }
  }
};

const Template = (args) => <Component {...args} />;

export const Avatar = Template.bind({});
Avatar.args = {
  src: "https://www.gravatar.com/avatar/205e460b479e2e5b48aec07710c08d50",
  userName: "Gratav User",
  size: "small"
};
