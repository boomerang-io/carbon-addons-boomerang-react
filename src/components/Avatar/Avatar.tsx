import React, { useState } from "react";
import { prefix } from "../../internal/settings";

interface AvatarProps {
  className?: string;
  size?: "small" | "medium" | "large";
  src: string;
  style?: React.CSSProperties;
  userName?: string;
}

const UserIcon = (props: React.SVGProps<SVGAElement> & { description?: string; userName: string }) => {
  const { description, userName, ...rest } = props;
  return (
    // @ts-expect-error TS(2322): Type '{ children: ("" | Element | undefined)[]; cl... Remove this comment to see the full error message
    <svg
      version="1.1"
      id="Layer_1"
      xmlns="http://www.w3.org/2000/svg"
      x="0px"
      y="0px"
      width="16px"
      height="16px"
      viewBox="0 0 16 16"
      role="img"
      {...rest}
    >
      <title>{`User icon for ${userName ?? "user"}`}</title>
      {description && <desc>{description}</desc>}
      <g id="user">
        <path
          id="Combined-Shape"
          fill="currentColor"
          d="M4,13.7C5.1,14.5,6.5,15,8,15s2.9-0.5,4-1.3v-1.2c0-0.8-0.7-1.5-1.5-1.5h-5C4.7,11,4,11.7,4,12.5V13.7
		z M3,12.9v-0.4C3,11.1,4.1,10,5.5,10h5c1.4,0,2.5,1.1,2.5,2.5v0.4c1.2-1.3,2-3,2-4.9c0-3.9-3.1-7-7-7S1,4.1,1,8
		C1,9.9,1.8,11.6,3,12.9z M8,16c-4.4,0-8-3.6-8-8s3.6-8,8-8s8,3.6,8,8S12.4,16,8,16z"
        />
        <path
          id="Oval-17"
          fill="currentColor"
          d="M8,8c1.1,0,2-0.9,2-2S9.1,4,8,4S6,4.9,6,6S6.9,8,8,8z M8,9C6.3,9,5,7.7,5,6s1.3-3,3-3s3,1.3,3,3S9.7,9,8,9z"
        />
      </g>
    </svg>
  );
};

function Avatar({ className = `${prefix}--bmrg-avatar`, size = "medium", src, style, userName, ...rest }: AvatarProps) {
  const [error, setError] = useState(false);

  const altText = `Avatar for ${userName}`;

  return error || !src ? (
    // @ts-expect-error TS(2741): Property 'userName' is missing in type '{ classNam... Remove this comment to see the full error message
    <UserIcon className={`${className} --${size || ""}`} description={altText} style={style} />
  ) : (
    <img
      alt={altText}
      className={`${className} --${size || ""}`}
      src={src}
      style={style}
      title={altText}
      onError={() => setError(true)}
      {...rest}
    />
  );
}

export default Avatar;
