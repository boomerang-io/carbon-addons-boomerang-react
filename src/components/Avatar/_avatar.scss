@use "../../global/config" as *;
@use "../../global/utils";

.#{$prefix}--bmrg-avatar {
  border-radius: 50%;

  &.--small {
    height: utils.rem(16px);
    width: utils.rem(16px);
    min-height: utils.rem(16px);
    min-width: utils.rem(16px);
  }
  &.--medium {
    height: utils.rem(32px);
    width: utils.rem(32px);
    min-height: utils.rem(32px);
    min-width: utils.rem(32px);
  }
  &.--large {
    height: utils.rem(64px);
    width: utils.rem(64px);
    min-height: utils.rem(64px);
    min-width: utils.rem(64px);
  }
}
