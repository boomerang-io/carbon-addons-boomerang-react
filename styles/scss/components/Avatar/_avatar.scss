@include exports('addons-bmrg-avatar') {
  .#{$prefix}--bmrg-avatar {
    border-radius: 50%;

    &.--small {
      height: rem(16px);
      width: rem(16px);
      min-height: rem(16px);
      min-width: rem(16px);
    }
    &.--medium {
      height: rem(32px);
      width: rem(32px);
      min-height: rem(32px);
      min-width: rem(32px);
    }
    &.--large {
      height: rem(64px);
      width: rem(64px);
      min-height: rem(64px);
      min-width: rem(64px);
    }
  }
}
