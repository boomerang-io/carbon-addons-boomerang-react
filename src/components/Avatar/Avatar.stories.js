import React from 'react';
import { storiesOf } from '@storybook/react';
import { text } from '@storybook/addon-knobs/react';

import Avatar from './Avatar';

const styleProp = { backgroundColor: '#50565b' };

storiesOf('Avatar', module).add('default', () => {
  return (
    <div style={styleProp}>
      <Avatar
        src={text('src', 'https://www.gravatar.com/avatar/205e460b479e2e5b48aec07710c08d50"')}
        userName={text('userName', 'Gratav User')}
      />
    </div>
  );
});
