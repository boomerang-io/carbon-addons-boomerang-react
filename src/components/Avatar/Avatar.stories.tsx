import React from "react";
import Avatar from "./Avatar";

export default {
  title: "Components/Avatar",
  components: Avatar,
  parameters: {
    docs: {
      description: {
        component: "A image component to display a user's profile image with a fallback user icon.",
      },
    },
  },
  argTypes: {
    size: {
      description: "Control the size",
      control: "select",
      options: ["small", "medium", "large"],
    },
  },
};

export const Default = (args) => <Avatar {...args} />;

Default.args = {
  src: "https://www.gravatar.com/avatar/205e460b479e2e5b48aec07710c08d50",
  userName: "Rick Deckard",
};
