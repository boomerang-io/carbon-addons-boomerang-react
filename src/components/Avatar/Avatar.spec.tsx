import React from "react";
import { expect, test } from "vitest";
import { render, screen } from "@testing-library/react";
import { axe } from "jest-axe";

import Avatar from "./Avatar";

describe("Avatar", () => {
  test("snapshot", () => {
    const { baseElement } = render(<Avatar src="https://www.gravatar.com/avatar/205e460b479e2e5b48aec07710c08d50" userName="Rick Deckard" />);
    expect(baseElement).toMatchSnapshot();
  });

  test("functional", async () => {
    render(<Avatar src="https://www.gravatar.com/avatar/205e460b479e2e5b48aec07710c08d50" userName="Rick Deckard" />);
    expect(await screen.findByAltText("Avatar for Rick Deckard")).toBeInTheDocument();
  });

  test("a11y", async () => {
    const { container } = render(<Avatar src="https://www.gravatar.com/avatar/205e460b479e2e5b48aec07710c08d50" userName="Rick Deckard" />);
    const results = await axe(container);
    expect(results).toHaveNoViolations();
  });
});
