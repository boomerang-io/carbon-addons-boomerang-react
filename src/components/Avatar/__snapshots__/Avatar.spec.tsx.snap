// Vitest Snapshot v1

exports[`Avatar > snapshot 1`] = `
<body>
  <div>
    <img
      alt="Avatar for Rick Deckard"
      class="cds--bmrg-avatar --medium"
      src="https://www.gravatar.com/avatar/205e460b479e2e5b48aec07710c08d50"
      title="Avatar for Rick Deckard"
    />
  </div>
</body>
`;
